flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp

```diff
@@ -289,7 +290,29 @@ mlir::LLVM::DITypeAttr DebugTypeGenerator::convertCharacterType(
     sizeInBits =
         charTy.getLen() * kindMapping.getCharacterBitsize(charTy.getFKind());
   } else {
-    return genPlaceholderType(context);
+    // In assumed length string, the len of the character is not part of the
+    // type but can be found at the runtime. Here we create an artificial
+    // variable that will contain that length. This variable is used as
+    // 'stringLength' in DIStringTypeAttr.
+    if (declOp && !declOp.getTypeparams().empty()) {
+      mlir::Operation *op = declOp.getTypeparams()[0].getDefiningOp();
+      if (auto unbox = mlir::dyn_cast_or_null<fir::UnboxCharOp>(op)) {
```

Comment:
Do you even need to restrict this to unbox?

I think you could just use `declOp.getTypeparams()[0]` directly and this would cover cases where there are no unbox like:

```
subroutine foo(n)
  integer(8) :: n
  character(len=n) :: c
end subroutine
```